File: src/main/java/de/hinundhergestellt/jhuh/shopify/ShopifyProductClient.java
package de.hinundhergestellt.jhuh.shopify;

import com.netflix.graphql.dgs.client.GraphQLClient;
import com.netflix.graphql.dgs.client.codegen.GraphQLQueryRequest;
import com.shopify.admin.client.ProductCreateGraphQLQuery;
import com.shopify.admin.client.ProductCreateProjectionRoot;
import com.shopify.admin.client.ProductsGraphQLQuery;
import com.shopify.admin.client.ProductsProjectionRoot;
import com.shopify.admin.types.PageInfo;
import com.shopify.admin.types.Product;
import com.shopify.admin.types.ProductConnection;
import com.shopify.admin.types.ProductCreateInput;
import com.shopify.admin.types.ProductCreatePayload;
import com.shopify.admin.types.ProductEdge;
import org.apache.commons.lang3.tuple.Pair;
import org.springframework.stereotype.Component;

import java.util.stream.Stream;

@Component
public class ShopifyProductClient {

    private final GraphQLClient apiClient;

    public ShopifyProductClient(GraphQLClient shopifyApiClient) {
        this.apiClient = shopifyApiClient;
    }

    public Stream<Product> findAll() {
        return PagingIterator.stream(this::findNextPage);
    }

    public void save(Product product) {
        if (product.getId() != null) {
            throw new IllegalArgumentException("Product id must be null");
        }

        var productInput = new ProductCreateInput();
        productInput.setTitle(product.getTitle());
        productInput.setVendor(product.getVendor());
        productInput.setProductType(product.getProductType());
        productInput.setTags(product.getTags());

        var query = ProductCreateGraphQLQuery.newRequest()
                .product(productInput)
                .build();

        // @formatter:off
        var root = new ProductCreateProjectionRoot<>()
                .product()
                    .id()
                    .parent()
                .userErrors()
                    .message()
                    .field();
        // @formatter:on

        var request = new GraphQLQueryRequest(query, root);
        var response = apiClient.executeQuery(request.serialize());
        var payload = response.extractValueAsObject("productCreate", ProductCreatePayload.class);
        if (!payload.getUserErrors().isEmpty()) {
            throw new IllegalArgumentException("Product creation failed: " + payload.getUserErrors());
        }
        product.setId(payload.getProduct().getId());
    }

    private Pair<Stream<Product>, PageInfo> findNextPage(String after) {
        var query = ProductsGraphQLQuery.newRequest()
                .first(100)
                .after(after)
                .build();

        var root = new ProductsProjectionRoot<>();
        var edgeProjection = root.edges();
        var productProjection = edgeProjection.node();
        productProjection.handle();
        productProjection.id();
        productProjection.title();
        productProjection.status();
        productProjection.tags();
        productProjection.vendor();

        var variantConnectionProjection = productProjection.variants(100, null, null, null, null, null);
        var variantEdgeProjection = variantConnectionProjection.edges();
        var variantProjection = variantEdgeProjection.node();
        variantProjection.barcode();
        variantProjection.compareAtPrice();
        variantProjection.id();
        variantProjection.price();
        variantProjection.sku();
        var inventoryItemProjection = variantProjection.inventoryItem();
        inventoryItemProjection.id();

        var pageInfoProjection = root.pageInfo();
        pageInfoProjection.endCursor();
        pageInfoProjection.hasNextPage();
        pageInfoProjection.startCursor();

        var request = new GraphQLQueryRequest(query, root);
        var response = apiClient.executeQuery(request.serialize());
        var payload = response.extractValueAsObject("products", ProductConnection.class);
        return Pair.ofNonNull(payload.getEdges().stream().map(ProductEdge::getNode), payload.getPageInfo());
    }
}

File: src/main/java/de/hinundhergestellt/jhuh/vendors/shopify/client/ShopifyProductClient.kt
package de.hinundhergestellt.jhuh.vendors.shopify.client

import com.netflix.graphql.dgs.client.GraphQLClient
import com.netflix.graphql.dgs.client.codegen.BaseSubProjectionNode
import com.netflix.graphql.dgs.client.codegen.GraphQLQueryRequest
import com.shopify.admin.client.ProductCreateGraphQLQuery
import com.shopify.admin.client.ProductCreateProjectionRoot
import com.shopify.admin.client.ProductDeleteGraphQLQuery
import com.shopify.admin.client.ProductDeleteProjectionRoot
import com.shopify.admin.client.ProductUpdateGraphQLQuery
import com.shopify.admin.client.ProductUpdateProjectionRoot
import com.shopify.admin.client.ProductsGraphQLQuery
import com.shopify.admin.client.ProductsProjectionRoot
import com.shopify.admin.types.PageInfo
import com.shopify.admin.types.ProductConnection
import com.shopify.admin.types.ProductCreatePayload
import com.shopify.admin.types.ProductDeletePayload
import com.shopify.admin.types.ProductUpdatePayload
import org.springframework.stereotype.Component

@Component
class ShopifyProductClient(
    private val apiClient: GraphQLClient
) {
    fun findAll() = pageAll {
        findAll(it)
    }

    fun create(product: UnsavedShopifyProduct): ShopifyProduct {
        val query = ProductCreateGraphQLQuery.newRequest()
            .product(product.toProductCreateInput())
            .build()

        // @formatter:off
        val root = ProductCreateProjectionRoot<BaseSubProjectionNode<*, *>, BaseSubProjectionNode<*, *>>()
            .product()
                .id()
                .options()
                    .id()
                    .parent()
                .parent()
            .userErrors()
                .message()
                .field()
        // @formatter:on

        val payload = apiClient.executeMutation(query, root, ProductCreatePayload::getUserErrors)
        val options = product.options.asSequence()
            .zip(payload.product.options.asSequence())
            .map { (option, created) -> ShopifyProductOption(option, created.id) }
            .toMutableList()
        return ShopifyProduct(product, payload.product.id, options)
    }

    fun update(product: ShopifyProduct) {
        val query = ProductUpdateGraphQLQuery.newRequest()
            .product(product.toProductUpdateInput())
            .build()

        // @formatter:off
        val root = ProductUpdateProjectionRoot<BaseSubProjectionNode<*, *>, BaseSubProjectionNode<*, *>>()
            .userErrors()
                .message()
                .field()
        // @formatter:on

        apiClient.executeMutation(query, root, ProductUpdatePayload::getUserErrors)
    }

    fun delete(product: ShopifyProduct) {
        val query = ProductDeleteGraphQLQuery.newRequest()
            .input(product.toProductDeleteInput())
            .build()

        // @formatter:off
        val root = ProductDeleteProjectionRoot<BaseSubProjectionNode<*, *>, BaseSubProjectionNode<*, *>>()
            .userErrors()
                .message()
                .field()
        // @formatter:on

        apiClient.executeMutation(query, root, ProductDeletePayload::getUserErrors)
    }

    private fun findAll(after: String?): Pair<Sequence<ShopifyProduct>, PageInfo> {
        val query = ProductsGraphQLQuery.newRequest()
            .first(250)
            .after(after)
            .build()

        // @formatter:off
        val root = ProductsProjectionRoot<BaseSubProjectionNode<*, *>, BaseSubProjectionNode<*, *>>()
            .edges()
                .node()
                    .handle()
                    .id()
                    .title()
                    .vendor()
                    .productType()
                    .status()
                        .parent()
                    .tags()
                    .hasOnlyDefaultVariant()
                    .variants(250, null, null, null, null, null)
                        .edges()
                            .node()
                                .id()
                                .title()
                                .price()
                                .sku()
                                .barcode()
                                .selectedOptions()
                                    .name()
                                    .value()
                                    .parent()
                                .parent()
                            .parent()
                        .pageInfo()
                            .hasNextPage()
                            .parent()
                        .parent()
                    .options()
                        .id()
                        .name()
                        .values()
                        .parent()
                    .metafields(null, null, 100, null, null, null, null)
                        .edges()
                            .node()
                                .id()
                                .namespace()
                                .key()
                                .value()
                                .type()
                                .parent()
                            .parent()
                        .pageInfo()
                            .hasNextPage()
                            .endCursor()
                            .parent()
                        .parent()
                    .parent()
                .parent()
            .pageInfo()
                .hasNextPage()
                .endCursor()
        // @formatter:on

        val request = GraphQLQueryRequest(query, root)
        val response = apiClient.executeQuery(request.serialize())
        require(!response.hasErrors()) { "Product collections find failed: " + response.errors }
        val payload = response.extractValueAsObject("products", ProductConnection::class.java)
        return Pair(
            payload.edges.asSequence()
                .map { it.node }
                .map { ShopifyProduct(it) },
            payload.pageInfo
        )
    }
}
